### DISCUSSION TITLE
Apollo Server v4 Integration into Firebase Cloud Functions

### DISCUSSION THREAD

**Discussion by TangrisJones** [apollographql/apollo-server#7772] [General]
Hey all 👋 , just wanted to mention this solution I seem to have found on deploying an apollo server v4 on a firebase cloud function. 

Seems to work for me so far but it dosent seem to be documented anywhere....? I've found related work [here](https://github.com/apollo-server-integrations/apollo-server-integration-google-cloud-functions), but this solution seems to be an alternative to that.

The TLDR; we deploy an [express server](https://www.apollographql.com/docs/apollo-server/api/express-middleware/) on a firebase [onRequest function](https://firebase.google.com/docs/functions/http-events).

Heres the code:

```
import * as functions from "firebase-functions";
import cors from "cors";
import express from "express";
import { expressMiddleware } from "@apollo/server/express4";
import { ApolloServer } from "@apollo/server";
import { schema } from "./schemas";
import http from "http";
import { ApolloServerPluginDrainHttpServer as apolloServerPluginDrainHttpServer } from "@apollo/server/plugin/drainHttpServer";

export interface Context {
  someField: string | null;
}

// Required logic for integrating with Express
export const app = express();
const httpServer = http.createServer(app);

export const server = new ApolloServer<Context>({
  schema,
  csrfPrevention: true,
  introspection: process.env.NODE_ENV === "development",
  plugins: [apolloServerPluginDrainHttpServer({ httpServer })],
});

server
  .start()
  .then(() =>
    // Set up our Express middleware to handle CORS, body parsing,
    // and our expressMiddleware function.
    app.use(
      "/",
      cors<cors.CorsRequest>(),
      express.json(),
      // expressMiddleware accepts the same arguments:
      // an Apollo Server instance and optional configuration options
      expressMiddleware(server, {
        context: async ({ req }) => ({ someField: "123" }),
      })
    )
  )
  .catch((err) => {
    throw err;
  });

export const api = functions.https.onRequest(app);
``` 

Seems to work well for me in initial testing! Let me know if anyone has thoughts / opinions on it... I'm a little nervous about this solution because of the lack of documentation... Hope it helps someone though 😄

**Comment by trevor-scheer** [comment]
One caveat here is you're not `await`ing `server.start()`, so it's possible that Apollo Server isn't ready to serve requests before it starts actually getting them. And in this configuration, you won't even have a `/` route established yet if the `start()` promise hasn't resolved.

**Comment by MichelFloyd** [comment]
[Here's a working repo](https://github.com/MichelFloyd/firebase-graphql)
